accepted/0000-better-cobbler-tftp-sync.md

```diff
@@ -0,0 +1,77 @@
+
+# Detailed design
+[design]: #detailed-design
+
+The solution of pushing the update files to proxies does not scale well. Especially with large installations and many proxies the time needed for a full sync grows linearly.
+
+The idea is to skip this step completely and move the logic to the proxy. In order to do that, we would need a TFTP server that is capable of dynamically checking if requested files are up-to-date on the file system, or if files need to be fetched from Uyuni. Also certain files need to be modified on the fly, since some files include the hostname of the Uyuni server.
+
+```
+      +----------------+
+      | Client Machine |
+      +-------+--------+
+              |
+            TFTP-
+           Request
+              |
+              v
+         +----+------+
+         |           |
+         |   TFTP    |
+         |           |
+         +-+-------+-+
+           |       |
+
+     Check local disk cache
+      for files. Otherwise
+     pull from Uyuni server.
+           |       |
+           |       |
+           v       v
++----------+-+   +-+------+
+| Local Disk |   | Uyuni  |
+| Cache      |   | Server |
++------------+   +--------+
+```
+
+The scenario would look like this. Once a client requests a file, the TFTP server would check if this file is in the local disk cache. If it's not, then it's immediately fetched via HTTP from Uyuni Server. If the file is present in the local disk cache, then the TFTP server would send a HEAD request to Uyuni server to get the current etag of the file. If the etags are matching, then the local file can be used. If not, then it needs to be fetched from Uyuni server and put into the local disk cache.
+For some files the Uyuni hostname needs to be replaced; this would be done right before delivering it to the client since etags wouldn't match otherwise.
+
+Ideally a proxy would be "warmed up" after a few client requests, since most of the files are then cached on disk. HEAD requests are releatively cheap and reasonably fast. So the first clients would encounter a slight delay, but all of the following requests for profiles with the same kernel and initrd would have nearly the same speed as the current implementation.
+
+Dealing with the thundering herd problem. Since it would be possible to have 300 clients requesting the same not yet synced, or outdated kernel, there needs to be a lock of some sorts on file requests. If there is a request to Uyuni server already running for the same file, there is no additional request allowed. Instead we'd just wait for the initial request to finish.
+
+One possible solution for the TFTP server would be [fbtftp](https://github.com/facebook/fbtftp) from Facebook. They ran into pretty much the same problem and created this dynamic TFTP server framework.
```

Comment:
Cache invalidation is in general a difficult problem, so indeed we should be careful before embarking into the 'roll your own solution' route.

`fbtftp` looks nice, are there any alternatives out there, specifically (ideally) caching proxies?

I suppose a true caching proxy is not really possible as TFTP does not have HTTP-like `Cache:` headers, correct?

Reply:
The thing is... In an ideal world we don't want to cache anything on this side because if the Cache is not correct (which is a situation which may occur), then the client would be provisioned with the wrong options. And yes TFTP does not have `Cache:` Headers afaik.

Reply:
Splitting the architecture into "tftp to http proxy" and "cache" is helpful here. As already mentioned in RocketChat, using the Squid instance on the Proxy machine will provide for a local cache and since we're talking about serving "files" at the upstream server, standard caching mechanisms can be applied.
Looking at the typical content of the (upstream) TFTP server, you'll see either small files with individual configuration (which may change frequently and is each used by only a single PXE client) and large (up to "huge") image files, which will rarely change. Distinguishing between these two at the Squid level keeps the config file always coming from the upstream and the image files from the cache.
Of course, even the config files could be cached, but considering that the over-all workflow of client installation will typically require live access to the upstream server anyhow, and the rather small difference between checking for a new file version or just serving these <5k files, I opt for "no caching of config files".